tools/DevDiagnostics/DevHome.DevDiagnostics/Helpers/LoaderSnapAssistantTool.cs

```diff
@@ -0,0 +1,257 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using CommunityToolkit.Mvvm.Input;
+using DevHome.Common.Extensions;
+using DevHome.DevDiagnostics.Controls;
+using DevHome.DevDiagnostics.Models;
+using DevHome.DevDiagnostics.Services;
+using DevHome.DevDiagnostics.TelemetryEvents;
+using DevHome.Service;
+using DevHome.Telemetry;
+using Microsoft.Diagnostics.Tracing;
+using Microsoft.Diagnostics.Tracing.Session;
+using Microsoft.UI.Xaml;
+using Microsoft.Win32;
+
+namespace DevHome.DevDiagnostics.Helpers;
+
+public class LoaderSnapAssistantTool
+{
+    private const string LoaderSnapsIdentifyNoLogsTelemetryName = "LoaderSnapsIdentifyNoLogs";
+    private const string LoaderSnapsIdentifyLogsTelemetryName = "LoaderSnapsIdentifyLogs";
+
+    private const string WindowsImageETWProvider = "2cb15d1d-5fc1-11d2-abe1-00a0c911f518"; /*EP_Microsoft-Windows-ImageLoad*/
+    private const uint LoaderSnapsFlag = 0x80; /* ETW_UMGL_LDR_SNAPS_FLAG */
+    private const string LoaderSnapsETWOpCode = "Opcode(215)";
+    private const string LoaderSnapsETWErrorLine = "LdrpProcessWork - ERROR: Unable to load";
+
+    private readonly DDInsightsService _insightsService;
+    private readonly IDevHomeService _devHomeService;
+    private readonly Microsoft.UI.Dispatching.DispatcherQueue _dispatcher;
+
+    private struct LoaderSnapFailure
+    {
+        public int Pid { get; set; }
+
+        public string? ErrorLog { get; set; }
+
+        public string? ImageName { get; set; }
+    }
+
+    private readonly Dictionary<int, LoaderSnapFailure> _loaderSnapFailures = new();
+
+    [Flags]
+    private enum TraceFlags
+    {
+        HeapTracing = 1,
+        CritSecTracing = 2,
+        LoaderSnaps = 4,
+    }
+
+    public LoaderSnapAssistantTool()
+    {
+        _dispatcher = Microsoft.UI.Dispatching.DispatcherQueue.GetForCurrentThread();
```

Comment:
I know DD does something different than the rest of Dev Home where apparently it's guaranteed to be running on the UI thread more than the rest can guarantee, but are you sure you're always going to be running on the UI thread here? If you're not, this returns null and you get in trouble.

Reply:
I added an assert here to confirm this is created on the UI thread.